added lives functionality

Eggsy07 opened the pull request:
1. The player has a total of 3 lives, represented by hearts (red for filled and white for empty).

2.When the player dies (explodes) in a certain level, he loses a life and remains in the same level as long as he still has available lives left. 

3.When the player dies while no more lives are available, the player returns to the previous level they were in, that means he gains one live to continue playing the game.

4.The player gains a life whenever he kills an enemy.



https://github.com/GamesForGood/CandyWrapperDeluxe/assets/88040864/07917f05-373f-41bc-84a7-4d16374b3941

Eggsy07 committed:
added lives functionality

Eggsy07 committed:
unit tests

Above/Bellow I don't exactly from where my description appears, but it is about unit tests, testing the way we increase or decrease lives on game player. I hope you like

Eggsy07 committed:
update

Filaipus commented on Script/BaseKine.gd:
Eu não adicionaria as mudanças deste ficheiro. Não tem nada a ver com a tua tarefa, é melhor só mandar para o git um Pull Request com ficheiros pertinentes á tarefa.

Filaipus commented on Script/Global.gd:
```diff
@@ -2,8 +2,20 @@ extends Node2D
\ No newline at end of file
+var Game
+
+func ModifyLives(var live, var action):
```
Esta função não deveria estar aqui, mas dentro da logica do Game. Global é para variáveis globais.

E eu recomendava modificares a função. Em vez de ter uma "action", devias permitir á função ser chamada só com um int. 

Por exemplo:

```
func ModifyLives(var live):

lives += live

if(lives < max_lives):
lives = max_lives

if(lives <= 1):
lives = 1
```

Filaipus commented on Script/Player.gd:
```diff
@@ -57,7 +57,7 @@ func DOHUD(arg : int):
 
 func _physics_process(delta):
-	# gravity
+	
```
Porque retiraste este comentario?

Eggsy07 replied:
Resolvido

Filaipus commented on Script/Player.gd:
```diff
@@ -115,9 +114,9 @@ func Explode(arg : Vector2, explosionType):
-	Explode(position, PLAYER_EXPLOSION.instance())
-	global.Game.Lose()
+		queue_free()
+		Explode(position, PLAYER_EXPLOSION.instance())
```
Porque esta mudança?

Filaipus commented on Script/btn.gd:
Eu não adicionaria as mudanças deste ficheiro. Não tem nada a ver com a tua tarefa, é melhor só mandar para o git um Pull Request com ficheiros pertinentes á tarefa.

Eggsy07 replied:
o ficheiro não sofreu alteração alguma foi se por acaso visitado pra fins de consulta nada mais.

Filaipus replied:
@Eggsy07 se não sofreu alteração, então não faças commit do ficheiro. Neste momento o teu Pull Request diz que modificaste 19 ficheiros, mas muitos dos ficheiros não tem a ver com a tua tarefa. Eu esperava só ver ficheiros pertinentes á tarefa neste Pull Request :)

Filaipus commented on addons/gut/gui/BottomPanelShortcuts.tscn:
Eu não adicionaria as mudanças deste ficheiro. Não tem nada a ver com a tua tarefa, é melhor só mandar para o git um Pull Request com ficheiros pertinentes á tarefa.

Eggsy07 replied:
o ficheiro não sofreu alteração alguma foi se por acaso visitado pra fins de consulta nada mais.

Filaipus commented on addons/gut/gui/GutBottomPanel.tscn:
Eu não adicionaria as mudanças deste ficheiro. Não tem nada a ver com a tua tarefa, é melhor só mandar para o git um Pull Request com ficheiros pertinentes á tarefa.

Eggsy07 replied:
o ficheiro não sofreu alteração alguma foi se por acaso visitado pra fins de consulta nada mais.

Filaipus commented on addons/gut/gui/OutputText.tscn:
Eu não adicionaria as mudanças deste ficheiro. Não tem nada a ver com a tua tarefa, é melhor só mandar para o git um Pull Request com ficheiros pertinentes á tarefa.

Eggsy07 replied:
o ficheiro não sofreu alteração alguma foi se por acaso visitado pra fins de consulta nada mais.

Filaipus commented on test/integration/test_Player.gd:
Eu não adicionaria as mudanças deste ficheiro. Não tem nada a ver com a tua tarefa, é melhor só mandar para o git um Pull Request com ficheiros pertinentes á tarefa.

Eggsy07 replied:
o ficheiro não sofreu alteração alguma foi se por acaso visitado pra fins de consulta nada mais.

Filaipus commented on test/unit/test_foobar.gd:
Eu não adicionaria as mudanças deste ficheiro. Não tem nada a ver com a tua tarefa, é melhor só mandar para o git um Pull Request com ficheiros pertinentes á tarefa.

Eggsy07 replied:
o ficheiro não sofreu alteração alguma foi se por acaso visitado pra fins de consulta nada mais.

Filaipus reviewed (changes requested):
Tens dois bugs neste momento. 
Quando o Jogador só tem 1 vida, ele não esta a voltar atrás no nível. Está a ficar no mesmo nível. Queremos que ele volte 1 nível atrás (mas só com uma vida).

O outro problema é que se chegares ao nível final, o jogo quebra. Não sei exatamente porque, mas temos que concertar isso visto que não era um problema antes. (vês a importância de testes? Se tivéssemos um teste para verificar que mudar de nível funciona direito, nunca terias que te preocupar com isto. Infelizmente não temos, mas um dia chegaremos lá!)

Eggsy07 committed:
update

basically I made changes on Global.gd and I just removed de method ModifyLives to Game.gd Script I also change it parameters to receive one value to set our player's life. Finally I had to update my test_Game.gd. I hope you like.

Eggsy07 committed:
update

now when the player got just one life he return to the previous level.

Eggsy07 committed:
Update Game.gd

Eggsy07 committed:
created HUD.gd

this script control the hearts panel  and the Game.tscn scene was automatically updated i've added a Panel and 2 TextureRect

Eggsy07 committed:
created test_Game.gd

Eggsy07 committed:
Update test_Game.gd

for my unit tests i just created test_Game.gd, where I tested lives increment, decrements, lives bellow zero and life over 3. And everything passed.

Eggsy07 committed:
Update Global.gd

Eggsy07 committed:
Update Global.gd

Here a I created my global variables for control how many lives my player has got on the game panel control

the player could have just 3 lives per play.
when he lose game while he still has lives available he remains to the same level e is at.
when he lose game while he has no lives available he return to the previous level he was with one life left.

when he win he gains one life and it will be automatically added to the life panel if he is missing some life.

lives can't be more than 3 and less than one

Eggsy07 committed:
Update Game.gd

Eggsy07 committed:
Update Game.gd

here firstly I made my panel, heartsempty and heartsfull visible on click  of key command to start the game.

on Lose() function I just added the code for verify if lives is equals to zero in other to set it again to one that is the our minimal condition. So if lives is over zero we keep our player playing the same level.

When the player win, I mean: on win() function I just call the method to increase one life on the player

at least I created the ModifyLives(int) function that increment or decrement lives according to the player situation call it whenever I need to increase or decrease life on the player and it works as hoped.

Eggsy07 committed:
Update Player.gd

Eggsy07 committed:
Update Player.gd

anFloor: the player is going to die() so we must decrement one the way I made it was to call the ModifyLives(int) Method.

Eggsy07 committed:
update
